File: solutions/javascript/merge-overlapping-intervals.js
class Pair { 
  constructor(first, second) { 
    this.first = first; 
    this.second = second; 
   } 
}

let mergeIntervals = function(v) {
  if (!v || v.length === 0) {
    return;
  }

  let result = [];
  result.push(new Pair(v[0].first, v[0].second));
  
  for (let i = 0; i < v.length; i++) {
    let x1 = v[i].first;
    let y1 = v[i].second;
    let x2 = result[result.length - 1].first;
    let y2 = result[result.length - 1].second;

    if (y2 >= x1) {
      result[result.length - 1].second = Math.max(y1, y2);
    } else {
      result.push(new Pair(x1, y1));
    }
  }
  return result;
};

let v = [new Pair(1, 5), new Pair(3, 7), new Pair(4, 6), 
         new Pair(6, 8), new Pair(10, 12), new Pair(11, 15)];

let result = mergeIntervals(v)

let result_str = ""
for (let i = 0; i < result.length; i++){
  result_str = result_str + "[" + result[i].first + ", " + result[i].second + "] "
}
console.log(result_str)

File: solutions/python/merge-overlapping-intervals.py
class Pair:
  def __init__(self, first, second):
    self.first = first
    self.second = second

def merge_intervals(v):
  if v == None or len(v) == 0 :
    return None

  result = []
  result.append(Pair(v[0].first, v[0].second))

  for i in range(1, len(v)):
    x1 = v[i].first
    y1 = v[i].second
    x2 = result[len(result) - 1].first
    y2 = result[len(result) - 1].second

    if y2 >= x1:
      result[len(result) - 1].second = max(y1, y2)
    else:
      result.append(Pair(x1, y1))
  return result;

v = [Pair(1, 5), Pair(3, 1), Pair(4, 6), 
     Pair(6, 8), Pair(10, 12), Pair(11, 15)]

result = merge_intervals(v)

for i in range(len(result)):
  print("[" + str(result[i].first) + ", " + str(result[i].second) + "]", end =" ")
